File: JS/LeetC/g75/linkedListCycle.js
/* ListNode = { val: val, next: ListNode | null }
  *
  */

function hasCycle(head) {
  if (head === null) {
    return false;
  }

  let slow = head;
  let fast = head.next;

  while (fast && fast.next && fast.next.next) {
    if (fast === slow) {
      return true;
    }

    slow = slow.next;
    fast = fast.next.next;
  }

  return false;
}

File: Java/hellointerview/05-linkedlist/linkedListCycle.java
public class Solution {
  public Boolean hasCycle(ListNode head) {
    ListNode slow = null;
    ListNode fast = head;

    while (fast != null && fast.next != null) {
      fast = fast.next.next;
      slow = slow.next;

      if (fast.equals(slow)) {
        return true;
      }
    }
    return false;
  }

}
